### PR TITLE
Event-based triggering with Aviso

### PR DESCRIPTION
**Pull request by ninaburg**
## Purpose

- Added configuration files for using the  [Aviso Notification System on EWC](https://confluence.ecmwf.int/display/EWCLOUDKB/Aviso+Notification+System+on+EWC)
- Added SQLite database to keep track of what forecasts have been processed
- Upload output files in s3 output bucket
- Remove `services/` directory because mainly empty or did not fit its purpose in this case
- Bump meteodata-lab and update functions

### CODE REVIEW

File: flexprep/config/aviso/listener_diss.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      domain: g
+      stream: oper
+    triggers:
+      - type: command
+        working_dir: "$HOME"
+        command: >
+          docker run \
+            --mount type=bind,source=$HOME/.sqlite/,destination=/src/app-root/db/
+            --env-file .env \
+            493666016161.dkr.ecr.eu-central-2.amazonaws.com/numericalweatherpredictions/flexpart_ifs/flexprep:2408.8a89852bd0a76e1930ea53b992c5035423d91485 \
```
Comment (ninaburg) [inline flexprep/config/aviso/listener_diss.yaml:16]: The image tag will be updated because once this is merged, I will deploy it and get a new tag.

File: Dockerfile [hunk 2]
```diff
@@ -49,4 +50,4 @@ ENV VERSION=$VERSION
 # uid 1001 is defined in the base-container-images for this purpose
 USER 1001
 
-CMD ["python", "-m", "flexprep"]
+ENTRYPOINT ["python", "-m", "flexprep"]
```
Comment (cfkanesan) [inline Dockerfile:53]: Why do you override the entrypoint here?

File: flexprep/__main__.py [hunk 3]
```diff
@@ -1,6 +1,74 @@
...
+        forecast_ref_time_str = f"{forecast_year}{args.date}{args.time[:4]}"
+        forecast_ref_time = dt.strptime(forecast_ref_time_str, "%Y%m%d%H%M")
+        return IFSForecast(
+            forecast_ref_time=forecast_ref_time,
+            step=args.step,
+            key=Path(args.location).name,
+            processed=False,
+        )
+    except Exception as e:
+        logging.error(f"Error creating IFSForecast object: {e}")
```
Comment (cfkanesan) [inline flexprep/__main__.py]: - common practice is to define a module level logger object with the name of the module to allow more fine tuned configuration e.g. `logger = logging.getLogger(__name__)`
- the `error` method will not include the stack trace whereas the `exception` method will
- you can always include the stack trace by setting `exc_info=True`

File: flexprep/__main__.py [hunk 4]
```diff
@@ -1,6 +1,74 @@
...
+from flexprep.domain.data_model import IFSForecast
+from flexprep.domain.prepare_processing import launch_pre_processing
+
+
+def parse_arguments():
+    """Parse and return command-line arguments."""
+    parser = argparse.ArgumentParser(description="Parse metadata of new file received")
+    parser.add_argument("--step", type=int, required=True, help="Step argument")
+    parser.add_argument(
+        "--date", type=str, required=True, help="Date argument (YYYY-MM-DD)"
```
Comment (cfkanesan) [inline flexprep/__main__.py]: The help text implies that the year is present in the date but further down it is prepended to the argument string and the parsing format does not match the format given here.

File: flexprep/domain/db_utils.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    def __init__(self) -> None:
+        try:
+            """Establish a database connection."""
+            self.db_path = CONFIG.main.db_path
+            self.conn = sqlite3.connect(self.db_path)
+            logging.info("Connected to database.")
+            self._initialize_db()
+        except sqlite3.Error as e:
+            logging.error(f"An error occurred: {e}")
```
Comment (cfkanesan) [inline flexprep/domain/db_utils.py]: I assume that the intention is just to log the exception, a raise is needed to let it continue to bubble up
```suggestion
            logging.error(f"An error occurred: {e}")
            raise
```

File: flexprep/domain/db_utils.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        finally:
```
Comment (cfkanesan) [inline flexprep/domain/db_utils.py]: Perhaps `else` is more appropriate

File: flexprep/domain/db_utils.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        self, forecast_ref_time: dt, step: int, key: str
+    ) -> None:
+        """Update the 'processed' field of a specific item to True."""
+        try:
+            with self.conn:
+                result = self.conn.execute(
+                    """
+                    UPDATE uploaded
+                    SET processed = 1
+                    WHERE forecast_ref_time = ? AND step = ? AND key = ?
```
Comment (cfkanesan) [inline flexprep/domain/db_utils.py]: The uploaded table should have a unique constraint on these columns, ideally through a unique index to make sure that this query does not slow down as the table grows. Alternatively, the `IFSForecast` object could get a row_id attribute which could be used to query the rows directly. Being the table's primary key, it would be automatically indexed.

File: flexprep/domain/prepare_processing.py [hunk 8]
```diff
@@ -1,96 +1,65 @@
...
-def launch_pre_processing(objects: dict[str, typing.Any]) -> None:
-    obj_in_s3 = aggregate_s3_objects(objects)
-    for fcst_ref_time, group in groupby(
-        obj_in_s3, key=lambda x: x["forecast_ref_time"]
-    ):
-        files_per_run = list(group)
-        step_zero = [file for file in files_per_run if file["step"] == 0]
-        steps = [file["step"] for file in files_per_run]
+    # Get step zero items and all steps
+    step_zero_items = [item.to_dict() for item in items_in_table if item.step == 0]
```
Comment (cfkanesan) [inline flexprep/domain/prepare_processing.py:24]: I think it's good to have domain specific objects, it's too bad that it's being transferred to a dict here. This could be the object of a separate PR.

File: flexprep/domain/processing.py [hunk 9]
```diff
@@ -98,36 +79,67 @@ def _load_and_validate_data(
...
+            ) as output_file:
+                # Write data to the temporary file
+                with open(output_file.name, "wb") as fout:
+                    for name, field in ds_out.items():
+                        if field.attrs.get("v_coord") == "hybrid":
+                            logging.info(f"Writing GRIB fields to {output_file.name}")
+                            grib_decoder.save(field, fout)
+
+            # Upload the file to S3
+            S3client().upload_file(output_file.name, key=key)
```
Comment (cfkanesan) [inline flexprep/domain/processing.py]: If this line is moved into the context manager above, the `delete=False` can be removed. Otherwise, you will need to remove the tempfile at some point

**Review by cfkanesan — CHANGES_REQUESTED** [review]
I didn't try following the Aviso Setup Guide since I don't believe Ihave access to EWC but I have some comments on the other parts.

File: Dockerfile (hunk 2 above)
Comment (ninaburg) [reply Dockerfile:53]: Using `CMD`, I was unable to pass arguments (such as `--step "${request.step}`) to the `docker run` command that needs to be executed when a notification arrives, resulting in the command not being parsed correctly. Here again the command:
```
          docker run \
            --mount type=bind,source=$HOME/.sqlite/,destination=/src/app-root/db/
            --env-file .env \
            493666016161.dkr.ecr.eu-central-2.amazonaws.com/numericalweatherpredictions/flexpart_ifs/flexprep:2408.8a89852bd0a76e1930ea53b992c5035423d91485 \
            --step "${request.step}" \
            --date "${request.date}" \
            --time "${request.time}" \
            --location "${location}"
```
Is there a reason to prefer using `CMD`?

Comment (cfkanesan) [reply Dockerfile:53]: I see. CMD defines the default command when calling `docker run` without an overriding command. If you wish to override the command, you need to pass the full command. It's also possible to override the entrypoint by passing the `--entrypoint` option so this mainly comes down to how you want to be able to use the image.

File: flexprep/__main__.py (hunk 3 above)
Comment (ninaburg) [reply flexprep/__main__.py]: - I've now used the module level logger which is defined in `config/settings.yaml`
- I've replaced the `error `method with the `exception `method to include the backtrace.

File: flexprep/__main__.py [hunk 10]
```diff
@@ -1,6 +1,74 @@
...
+    )
+    parser.add_argument(
+        "--time", type=str, required=True, help="Time argument (HH:MM:SS)"
+    )
+    parser.add_argument("--location", type=str, required=True, help="Location argument")
```
Comment (victoria-cherkas) [inline flexprep/__main__.py:21]: What is location here?

File: flexprep/__main__.py (hunk 4 above)
Comment (ninaburg) [reply flexprep/__main__.py]: Right, changed it to `mmdd`. Same for `time`: `HHMM` (without seconds).

File: flexprep/__main__.py [hunk 11]
```diff
@@ -1,6 +1,74 @@
...
+def create_ifs_forecast_obj(args):
+    """Create an IFSForecast object based on the parsed arguments."""
+    try:
+        # Get the current year and month
+        now = dt.now()
+        current_year = now.year
+        current_month = now.month
+
+        # Extract month from args.date
+        forecast_month = int(args.date[:2])
```
Comment (victoria-cherkas) [inline flexprep/__main__.py:36]: Doesnt align with the datetime format in argument description

File: flexprep/__main__.py [hunk 12]
```diff
@@ -1,6 +1,74 @@
...
+        if current_month == 1 and forecast_month == 12:
+            # If current month is Jan and forecast month is Dec, use the prev year
+            forecast_year = current_year - 1
+        else:
+            # Otherwise, use the current year
+            forecast_year = current_year
+
+        # Combine date and time to create forecast_ref_time
+        forecast_ref_time_str = f"{forecast_year}{args.date}{args.time[:4]}"
+        forecast_ref_time = dt.strptime(forecast_ref_time_str, "%Y%m%d%H%M")
```
Comment (victoria-cherkas) [inline flexprep/__main__.py]: same here with time format in argument description

File: flexprep/config/aviso/listener_diss.yaml [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+            --step "${request.step}" \
+            --date "${request.date}" \
+            --time "${request.time}" \
+            --location "${location}"
+        environment:
+          STEP: "${request.step}"
+          TIME: "The time is ${request.time}"
```
Comment (victoria-cherkas) [inline flexprep/config/aviso/listener_diss.yaml]: Are these environment vars needed ? We should remove them if not

File: flexprep/domain/db_utils.py (hunk 5 above)
Comment (ninaburg) [reply flexprep/domain/db_utils.py]: I've now added a raise in every `except `block

File: flexprep/domain/db_utils.py (hunk 7 above)
Comment (ninaburg) [reply flexprep/domain/db_utils.py]: I introduced a row_id attribute to IFSForecast obj.

File: flexprep/config/aviso/listener_diss.yaml (hunk 13 above)
Comment (ninaburg) [reply flexprep/config/aviso/listener_diss.yaml]: No they are not needed. I removed them.

File: flexprep/__main__.py (hunk 10 above)
Comment (ninaburg) [reply flexprep/__main__.py:21]: `location` here is the absolute path in the S3. For example: 
"location": "s3://storage.ecmwf.europeanweather.cloud/ecpds/xxx.xx" from [Aviso Doc ](https://confluence.ecmwf.int/display/EWCLOUDKB/Aviso+Notification+System+on+EWC#AvisoNotificationSystemonEWC-Request). 
I then only take the key of the file from this location argument we obtain from Aviso:  `key=Path(args.location).name`

File: flexprep/__main__.py (hunk 11 above)
Comment (ninaburg) [reply flexprep/__main__.py:36]: I corrected the arg description which is now: `mmdd` and `[:2]` should be correct now

File: flexprep/__main__.py (hunk 12 above)
Comment (ninaburg) [reply flexprep/__main__.py]: I also corrected time in the arg. description which is now `HHMM` (without seconds) such that `[:4]` can be completely removed now.

File: flexprep/domain/prepare_processing.py (hunk 8 above)
Comment (ninaburg) [reply flexprep/domain/prepare_processing.py:24]: I'll address that in a separate PR.

**Review by victoria-cherkas — APPROVED** [review]

File: flexprep/domain/processing.py [hunk 14]
```diff
@@ -65,69 +50,96 @@ def _sort_and_download_files(
...
+                )
+                ds_in |= metadata.extract_pv(ds_in["u"].message)
+
+            return ds_in
+
+        except Exception as e:
+            _LOGGER.exception(f"Data loading and validation failed: {e}")
+            raise
+
+        else:
```
Comment (cfkanesan) [inline flexprep/domain/processing.py]: finally was correct here, I suppose you want to remove the files whether the process was successful or not

**Review by cfkanesan — APPROVED** [review]
